modules/cluster_estimation/cluster_estimation.py

```diff
@@ -0,0 +1,342 @@
+"""
+Take in bounding box coordinates from Geolocation and use to estimate landing pad locations.
+Returns an array of classes, each containing the x coordinate, y coordinate, and spherical 
+covariance of each landing pad estimation.
+"""
+
+import numpy as np
+import sklearn
+import sklearn.datasets
+import sklearn.mixture
+
+from .. import object_in_world
+from .. import detection_in_world
+
+
+class ClusterEstimation:
+    """
+    Estimate landing pad locations based on landing pad ground detection. Estimation
+    works by predicting 'cluster centers' from groups of closely placed landing pad
+    detections.
+
+    ATTRIBUTES
+    ----------
+    random_state: int
+        Seed for randomizer, to get consistent results.
+    
+    min_activation_threshold: int
+        Minimum total data points before model runs.
+
+    new_points_per_run: int
+        Minimum number of new data points that must be collected
+        before running model.
+
+    METHODS
+    -------
+    reset_model()
+        Unfits the model of any data.
+
+    run()
+        Take in list of landing pad detections and return list of estimated landing pad locations
+        if number of detections is sufficient, or if manually forced to run.
+
+    __decide_to_run()
+        Decide when to run cluster estimation model.
+
+    __sort_by_weights()
+        Sort input model output list by weights in descending order.
+
+    __get_distance()
+        Calculates distance between a point and a cluster center.
+
+    __convert_detections_to_point()
+        Convert DetectionInWorld input object to a [x,y] position to store.
+
+    __filter_by_points_ownership()
+        Removes any clusters that don't have any points belonging to it.
+
+    __filter_by_covariances()
+        Removes any cluster with covariances much higher than the lowest covariance value.
+    """
+    __create_key = object()
+
+    # VGMM Hyperparameters
+    __MODEL_INIT_PARAM = "k-means++"
+    __COVAR_TYPE = "spherical"
+    __WEIGHT_CONCENTRATION_PRIOR = 100
+    __MEAN_PRECISION_PRIOR = 1E-6
+    __MAX_MODEL_ITERATIONS = 1000
+
+    # Hyperparameters to clean up model outputs
+    __WEIGHT_DROP_THRESHOLD = 0.1
+    __MAX_COVARIANCE_THRESHOLD = 10
+
+    # Real-world scenario Hyperparameters
+    __MAX_NUM_COMPONENTS = 10  # assumed maximum number of real landing pads
+
+    @classmethod
+    def create(cls,
+               min_activation_threshold: int,
+               min_new_points_to_run: int,
+               random_state: int) -> "tuple[bool, ClusterEstimation | None]":
+        """
+        Data requirement conditions for estimation worker to run.
+        """
+        if min_activation_threshold < 0 or min_new_points_to_run < 0 or random_state < 0:
```

Comment:
Are the minimums allowed to be 0? What happens if there are no points for the library to use? What happens if there are no new points but `run()` is called anyway? What happens if `run()` is called with override but there are no new points (it might be worth storing `detections_in_world` instead of refitting, or it might be worth refitting anyway)?

Reply:
Note:
- any incoming data is stored into the current bucket regardless if the model is ran in that call or not

1. min_new_points_to_run can be = 0
- This minimum # of new points requirement becomes always true, even if incoming Detections list is empty. Worker now only checks if total # of points meet min_activation_threshold for decision.

2. min_activation_threshold must be 1 or greater
- can not be zero since model object will error if # points = 0 for the .fit() call

4. run_override = True but total # points stored = 0:
- model will error so this won’t be run, treated as if __decide_to_run() is false.
   
5. run_override = True but # of new points = 0:
 - model runs anyways if total # points > 0, and is re-fitted onto the same data.